rllib/algorithms/algorithm.py

```diff
@@ -311,40 +308,50 @@ def from_state(state: Dict) -> "Algorithm":
     @PublicAPI
     def __init__(
         self,
-        config: Optional[Union[PartialAlgorithmConfigDict, AlgorithmConfig]] = None,
-        env: Optional[Union[str, EnvType]] = None,
+        config: Union[AlgorithmConfig, PartialAlgorithmConfigDict],
+        env=None,  # deprecated
         logger_creator: Optional[Callable[[], Logger]] = None,
         **kwargs,
     ):
         """Initializes an Algorithm instance.
 
         Args:
-            config: Algorithm-specific configuration dict.
-            env: Name of the environment to use (e.g. a gym-registered str),
-                a full class path (e.g.
-                "ray.rllib.examples.env.random_env.RandomEnv"), or an Env
-                class directly. Note that this arg can also be specified via
-                the "env" key in `config`.
+            config: Algorithm-specific configuration object.
             logger_creator: Callable that creates a ray.tune.Logger
                 object. If unspecified, a default logger is created.
             **kwargs: Arguments passed to the Trainable base class.
 
         """
 
-        # User provided (partial) config (this may be w/o the default
-        # Algorithm's Config object). Will get merged with AlgorithmConfig()
-        # in self.setup().
-        config = config or {}
-        # Resolve AlgorithmConfig into a plain dict.
+        # Resolve possible dict into an AlgorithmConfig object.
         # TODO: In the future, only support AlgorithmConfig objects here.
-        if isinstance(config, AlgorithmConfig):
-            config = config.to_dict()
+        if isinstance(config, dict):
```

Comment:
Retro-support old-style python config dicts for a while.
We will translate these into AlgorithmConfig objects from here on, freeze the AlgorithmConfig object (so it cannot be changed anymore by anyone), and use it under the hood in all Algorithms.
The added backward-compat mechanism to look up old-style dict keys (str) from those AlgorithmConfig objects makes sure, this even works for custom algos that still think they are dealing with a config dict.